docs/ovs-pipeline.md

```diff
@@ -0,0 +1,561 @@
+going through the gateway. The Virtual MAC is used as the destination MAC
+address for all the traffic being tunnelled.
+
+If you dump the flows for this table, you may see the following:
+```
+1. table=20, priority=200,arp,arp_tpa=10.10.1.1,arp_op=1 actions=move:NXM_OF_ETH_SRC[]->NXM_OF_ETH_DST[],mod_dl_src:aa:bb:cc:dd:ee:ff,load:0x2->NXM_OF_ARP_OP[],move:NXM_NX_ARP_SHA[]->NXM_NX_ARP_THA[],load:0xaabbccddeeff->NXM_NX_ARP_SHA[],move:NXM_OF_ARP_SPA[]->NXM_OF_ARP_TPA[],load:0xa0a0101->NXM_OF_ARP_SPA[],IN_PORT
+2. table=20, priority=80,ip actions=drop
+3. table=20, priority=190,arp actions=NORMAL
+```
+
+Flow 1 is the "ARP responder" for the peer Node whose local Pod subnet is
+10.10.1.0/24. If we were to look at the routing table for the local Node, we
+would see the following "onlink" route:
+```
+10.10.1.0/24 via 10.10.1.1 dev gw0 onlink
+```
+A similar route is installed on the gateway (gw0) interface every time the
+Antrea Node Route Controller is notified that a new Node has joined the
+cluster. The route must be marked as "onlink" since the kernel does not have a
+route to the peer gateway 10.10.1.1: we trick the kernel into believing that
+10.10.1.1 is directly connected to the local Node, even though it is on the
+other side of the tunnel.
+
+Flow 3 ensures that OVS handle the remainder of ARP traffic as a regular L2
+learning switch (using the `normal` action). In particular, this takes care of
+forwarding ARP requests and replies between local Pods.
+
+Flow 2 drops all IP traffic; this flow should never be used since only ARP
+traffic is sent to this table (by [SpoofGuardTable]).
+
+### ConntrackTable (30)
+
+The sole purpose of this table is to invoke the `ct` action on all packets and
+set the `ct_zone` (connection tracking context) to an hard-coded value, then
+resubmit traffic to [ConntrackStateTable]. If you dump the flows for this table,
+you should only see 2 flows:
+```
+1. table=30, priority=80,ip actions=resubmit(,31)
+2. table=30, priority=200,ip actions=ct(table=31,zone=65520)
+```
+
+In practice, flow 1 should never be used.
+
+A `ct_zone` is simply used to isolate connection tracking rules. It is similar
+in spirit to the more generic Linux network namespaces, but `ct_zone` is
+specific to conntrack and has less overhead.
+
+After invoking the ct action, packets will be in the "tracked" (`trk`) state and
+all [connection tracking
+fields](http://www.openvswitch.org//support/dist-docs/ovs-fields.7.txt) will be
+set to the correct value. Packets will then move on to [ConntrackStateTable].
+
+Refer to [this
+document](http://docs.openvswitch.org/en/latest/tutorials/ovs-conntrack/) for
+more information on connection tracking in OVS.
+
+### ConntrackStateTable (31)
+
+This table handles all "tracked" packets (all packets are moved to the tracked
+state by the previous table, [ConntrackTable]). It serves the following
+purposes:
+ * keeps track of connections going through the gateway port; for all packets
```

Comment:
I think I got this wrong: this mechanism applies to reverse traffic from a tunnel as well, not just from local backend Pods. @wenyingd could you confirm?

Reply:
ConntrackStateTable  also commits packets from the tunnel port into the ct_zone. It works by flow "table=31, priority=190,ct_state=+new+trk,ip actions=ct(commit,table=40,zone=65520)"